Change SaveAsCopy's to fetch existing data with GraphQL

connorwinston opened the pull request:
## Scope

What's changed:

Using GraphQL to fetch existing item data when using the "Save As Copy" functionality, as opposed to the REST API before which could fail if too many fields had been selected (URL too long). 

## Potential Risks / Drawbacks

- Any differences between REST/GraphQL when it comes to fetching data
- New logic to merge item data

---

Fixes #24237

Fixes WEB-673

connorwinston committed:
Change non-system collection SaveAsCopy's to fetch existing data with GraphQL instead of rest.

connorwinston committed:
Prettier

connorwinston committed:
Create early-toes-rule.md

formfcw reviewed (dismissed):
Thank you @connorwinston ! ❤️ 

In my test I got this 400 graphql validation error:

<details><summary>show output</summary>

```json
{
    "errors": [
        {
            "message": "GraphQL validation error.",
            "extensions": {
                "errors": [
                    {
                        "message": "Fields \"item\" conflict because \"events\" and \"blogs\" are different fields. Use different aliases on the fields to fetch both if this was intentional.",
                        "locations": [
                            { "line": 1, "column": 51 },
                            { "line": 1, "column": 143 }
                        ]
                    },
                    {
                        "message": "Cannot query field \"events\" on type \"alignment_M2A\".",
                        "locations": [{ "line": 1, "column": 51 }]
                    },
                    {
                        "message": "Cannot query field \"blogs\" on type \"alignment_M2A\".",
                        "locations": [{ "line": 1, "column": 143 }]
                    },
                    {
                        "message": "Fields \"item\" conflict because \"calendar\" and \"blogs\" are different fields. Use different aliases on the fields to fetch both if this was intentional.",
                        "locations": [
                            { "line": 1, "column": 302 },
                            { "line": 1, "column": 360 }
                        ]
                    },
                    {
                        "message": "Fields \"item\" conflict because \"calendar\" and \"cards\" are different fields. Use different aliases on the fields to fetch both if this was intentional.",
                        "locations": [
                            { "line": 1, "column": 302 },
                            { "line": 1, "column": 1080 }
                        ]
                    },
                    {
                        "message": "Fields \"item\" conflict because \"calendar\" and \"status\" are different fields. Use different aliases on the fields to fetch both if this was intentional.",
                        "locations": [
                            { "line": 1, "column": 302 },
                            { "line": 1, "column": 1187 }
                        ]
                    },
                    {
                        "message": "Fields \"item\" conflict because \"calendar\" and \"alignment\" are different fields. Use different aliases on the fields to fetch both if this was intentional.",
                        "locations": [
                            { "line": 1, "column": 302 },
                            { "line": 1, "column": 1213 }
                        ]
                    },
                    {
                        "message": "Fields \"item\" conflict because \"calendar\" and \"posts\" are different fields. Use different aliases on the fields to fetch both if this was intentional.",
                        "locations": [
                            { "line": 1, "column": 302 },
                            { "line": 1, "column": 1470 }
                        ]
                    },
                    {
                        "message": "Fields \"item\" conflict because \"calendar\" and \"events\" are different fields. Use different aliases on the fields to fetch both if this was intentional.",
                        "locations": [
                            { "line": 1, "column": 302 },
                            { "line": 1, "column": 1577 }
                        ]
                    },
                    {
                        "message": "Fields \"item\" conflict because \"blogs\" and \"cards\" are different fields. Use different aliases on the fields to fetch both if this was intentional.",
                        "locations": [
                            { "line": 1, "column": 360 },
                            { "line": 1, "column": 1080 }
                        ]
                    },
                    {
                        "message": "Fields \"item\" conflict because \"blogs\" and \"status\" are different fields. Use different aliases on the fields to fetch both if this was intentional.",
                        "locations": [
                            { "line": 1, "column": 360 },
                            { "line": 1, "column": 1187 }
                        ]
                    },
                    {
                        "message": "Fields \"item\" conflict because \"blogs\" and \"alignment\" are different fields. Use different aliases on the fields to fetch both if this was intentional.",
                        "locations": [
                            { "line": 1, "column": 360 },
                            { "line": 1, "column": 1213 }
                        ]
                    },
                    {
                        "message": "Fields \"item\" conflict because \"blogs\" and \"posts\" are different fields. Use different aliases on the fields to fetch both if this was intentional.",
                        "locations": [
                            { "line": 1, "column": 360 },
                            { "line": 1, "column": 1470 }
                        ]
                    },
                    {
                        "message": "Fields \"item\" conflict because \"blogs\" and \"events\" are different fields. Use different aliases on the fields to fetch both if this was intentional.",
                        "locations": [
                            { "line": 1, "column": 360 },
                            { "line": 1, "column": 1577 }
                        ]
                    },
                    {
                        "message": "Fields \"item\" conflict because \"cards\" and \"status\" are different fields. Use different aliases on the fields to fetch both if this was intentional.",
                        "locations": [
                            { "line": 1, "column": 1080 },
                            { "line": 1, "column": 1187 }
                        ]
                    },
                    {
                        "message": "Fields \"item\" conflict because \"cards\" and \"alignment\" are different fields. Use different aliases on the fields to fetch both if this was intentional.",
                        "locations": [
                            { "line": 1, "column": 1080 },
                            { "line": 1, "column": 1213 }
                        ]
                    },
                    {
                        "message": "Fields \"item\" conflict because \"cards\" and \"posts\" are different fields. Use different aliases on the fields to fetch both if this was intentional.",
                        "locations": [
                            { "line": 1, "column": 1080 },
                            { "line": 1, "column": 1470 }
                        ]
                    },
                    {
                        "message": "Fields \"item\" conflict because \"cards\" and \"events\" are different fields. Use different aliases on the fields to fetch both if this was intentional.",
                        "locations": [
                            { "line": 1, "column": 1080 },
                            { "line": 1, "column": 1577 }
                        ]
                    },
                    {
                        "message": "Fields \"item\" conflict because \"status\" and \"alignment\" are different fields. Use different aliases on the fields to fetch both if this was intentional.",
                        "locations": [
                            { "line": 1, "column": 1187 },
                            { "line": 1, "column": 1213 }
                        ]
                    },
                    {
                        "message": "Fields \"item\" conflict because \"status\" and \"posts\" are different fields. Use different aliases on the fields to fetch both if this was intentional.",
                        "locations": [
                            { "line": 1, "column": 1187 },
                            { "line": 1, "column": 1470 }
                        ]
                    },
                    {
                        "message": "Fields \"item\" conflict because \"status\" and \"events\" are different fields. Use different aliases on the fields to fetch both if this was intentional.",
                        "locations": [
                            { "line": 1, "column": 1187 },
                            { "line": 1, "column": 1577 }
                        ]
                    },
                    {
                        "message": "Fields \"item\" conflict because \"alignment\" and \"posts\" are different fields. Use different aliases on the fields to fetch both if this was intentional.",
                        "locations": [
                            { "line": 1, "column": 1213 },
                            { "line": 1, "column": 1470 }
                        ]
                    },
                    {
                        "message": "Fields \"item\" conflict because \"alignment\" and \"events\" are different fields. Use different aliases on the fields to fetch both if this was intentional.",
                        "locations": [
                            { "line": 1, "column": 1213 },
                            { "line": 1, "column": 1577 }
                        ]
                    },
                    {
                        "message": "Fields \"item\" conflict because \"posts\" and \"events\" are different fields. Use different aliases on the fields to fetch both if this was intentional.",
                        "locations": [
                            { "line": 1, "column": 1470 },
                            { "line": 1, "column": 1577 }
                        ]
                    },
                    {
                        "message": "Cannot query field \"calendar\" on type \"alignment_M2A_2\".",
                        "locations": [{ "line": 1, "column": 302 }]
                    },
                    {
                        "message": "Cannot query field \"blogs\" on type \"alignment_M2A_2\".",
                        "locations": [{ "line": 1, "column": 360 }]
                    },
                    {
                        "message": "Cannot query field \"cards\" on type \"alignment_M2A_2\".",
                        "locations": [{ "line": 1, "column": 1080 }]
                    },
                    {
                        "message": "Cannot query field \"status\" on type \"alignment_M2A_2\".",
                        "locations": [{ "line": 1, "column": 1187 }]
                    },
                    {
                        "message": "Cannot query field \"alignment\" on type \"alignment_M2A_2\". Did you mean \"alignment_id\"?",
                        "locations": [{ "line": 1, "column": 1213 }]
                    },
                    {
                        "message": "Cannot query field \"posts\" on type \"alignment_M2A_2\". Did you mean \"sort\"?",
                        "locations": [{ "line": 1, "column": 1470 }]
                    },
                    {
                        "message": "Cannot query field \"events\" on type \"alignment_M2A_2\".",
                        "locations": [{ "line": 1, "column": 1577 }]
                    },
                    {
                        "message": "Field \"user_created\" of type \"directus_users\" must have a selection of subfields. Did you mean \"user_created { ... }\"?",
                        "locations": [{ "line": 1, "column": 1702 }]
                    },
                    {
                        "message": "Field \"user_updated\" of type \"directus_users\" must have a selection of subfields. Did you mean \"user_updated { ... }\"?",
                        "locations": [{ "line": 1, "column": 1715 }]
                    },
                    {
                        "message": "Field \"alignment_2_id\" of type \"alignment\" must have a selection of subfields. Did you mean \"alignment_2_id { ... }\"?",
                        "locations": [{ "line": 1, "column": 1763 }]
                    },
                    {
                        "message": "Field \"alignment_id\" of type \"alignment\" must have a selection of subfields. Did you mean \"alignment_id { ... }\"?",
                        "locations": [{ "line": 1, "column": 1778 }]
                    },
                    {
                        "message": "Field \"cards_id\" of type \"cards\" must have a selection of subfields. Did you mean \"cards_id { ... }\"?",
                        "locations": [{ "line": 1, "column": 1791 }]
                    },
                    {
                        "message": "Field \"kanban_o2m_id\" of type \"kanban_test\" must have a selection of subfields. Did you mean \"kanban_o2m_id { ... }\"?",
                        "locations": [{ "line": 1, "column": 1805 }]
                    },
                    {
                        "message": "Field \"post_id\" of type \"posts\" must have a selection of subfields. Did you mean \"post_id { ... }\"?",
                        "locations": [{ "line": 1, "column": 1819 }]
                    },
                    {
                        "message": "Field \"dates\" of type \"[event_dates]\" must have a selection of subfields. Did you mean \"dates { ... }\"?",
                        "locations": [{ "line": 1, "column": 1827 }]
                    },
                    {
                        "message": "Field \"user_created\" of type \"directus_users\" must have a selection of subfields. Did you mean \"user_created { ... }\"?",
                        "locations": [{ "line": 1, "column": 1861 }]
                    },
                    {
                        "message": "Field \"user_updated\" of type \"directus_users\" must have a selection of subfields. Did you mean \"user_updated { ... }\"?",
                        "locations": [{ "line": 1, "column": 1887 }]
                    },
                    {
                        "message": "Field \"treeview_id\" of type \"alignment\" must have a selection of subfields. Did you mean \"treeview_id { ... }\"?",
                        "locations": [{ "line": 1, "column": 1913 }]
                    },
                    {
                        "message": "Field \"treeview_2_id\" of type \"alignment\" must have a selection of subfields. Did you mean \"treeview_2_id { ... }\"?",
                        "locations": [{ "line": 1, "column": 1925 }]
                    },
                    {
                        "message": "Field \"files\" of type \"[alignment_files]\" must have a selection of subfields. Did you mean \"files { ... }\"?",
                        "locations": [{ "line": 1, "column": 1969 }]
                    },
                    {
                        "message": "Field \"files2\" of type \"[alignment_files_1]\" must have a selection of subfields. Did you mean \"files2 { ... }\"?",
                        "locations": [{ "line": 1, "column": 1975 }]
                    },
                    {
                        "message": "Field \"calendar_item\" of type \"calendar\" must have a selection of subfields. Did you mean \"calendar_item { ... }\"?",
                        "locations": [{ "line": 1, "column": 1982 }]
                    },
                    {
                        "message": "Field \"M2A\" of type \"[alignment_M2A]\" must have a selection of subfields. Did you mean \"M2A { ... }\"?",
                        "locations": [{ "line": 1, "column": 1996 }]
                    },
                    {
                        "message": "Field \"M2A_2\" of type \"[alignment_M2A_2]\" must have a selection of subfields. Did you mean \"M2A_2 { ... }\"?",
                        "locations": [{ "line": 1, "column": 2000 }]
                    },
                    {
                        "message": "Field \"m2m\" of type \"[alignment_events]\" must have a selection of subfields. Did you mean \"m2m { ... }\"?",
                        "locations": [{ "line": 1, "column": 2006 }]
                    },
                    {
                        "message": "Field \"m2m_2\" of type \"[alignment_events_1]\" must have a selection of subfields. Did you mean \"m2m_2 { ... }\"?",
                        "locations": [{ "line": 1, "column": 2010 }]
                    },
                    {
                        "message": "Field \"o2m\" of type \"[posts]\" must have a selection of subfields. Did you mean \"o2m { ... }\"?",
                        "locations": [{ "line": 1, "column": 2016 }]
                    },
                    {
                        "message": "Field \"o2m_2\" of type \"[events]\" must have a selection of subfields. Did you mean \"o2m_2 { ... }\"?",
                        "locations": [{ "line": 1, "column": 2020 }]
                    },
                    {
                        "message": "Field \"tv\" of type \"[alignment]\" must have a selection of subfields. Did you mean \"tv { ... }\"?",
                        "locations": [{ "line": 1, "column": 2026 }]
                    },
                    {
                        "message": "Field \"user_updated\" of type \"directus_users\" must have a selection of subfields. Did you mean \"user_updated { ... }\"?",
                        "locations": [{ "line": 1, "column": 2073 }]
                    },
                    {
                        "message": "Field \"user_created\" of type \"directus_users\" must have a selection of subfields. Did you mean \"user_created { ... }\"?",
                        "locations": [{ "line": 1, "column": 2099 }]
                    },
                    {
                        "message": "Field \"relation_sort_id\" of type \"relation_sort\" must have a selection of subfields. Did you mean \"relation_sort_id { ... }\"?",
                        "locations": [{ "line": 1, "column": 2112 }]
                    },
                    {
                        "message": "Field \"post_id\" of type \"posts\" must have a selection of subfields. Did you mean \"post_id { ... }\"?",
                        "locations": [{ "line": 1, "column": 2235 }]
                    },
                    {
                        "message": "Field \"dates\" of type \"[event_dates]\" must have a selection of subfields. Did you mean \"dates { ... }\"?",
                        "locations": [{ "line": 1, "column": 2273 }]
                    },
                    {
                        "message": "Field \"user_created\" of type \"directus_users\" must have a selection of subfields. Did you mean \"user_created { ... }\"?",
                        "locations": [{ "line": 1, "column": 2296 }]
                    },
                    {
                        "message": "Field \"user_updated\" of type \"directus_users\" must have a selection of subfields. Did you mean \"user_updated { ... }\"?",
                        "locations": [{ "line": 1, "column": 2322 }]
                    },
                    {
                        "message": "Field \"calendar\" of type \"calendar\" must have a selection of subfields. Did you mean \"calendar { ... }\"?",
                        "locations": [{ "line": 1, "column": 2363 }]
                    },
                    {
                        "message": "Field \"m2m_test\" of type \"[posts_posts]\" must have a selection of subfields. Did you mean \"m2m_test { ... }\"?",
                        "locations": [{ "line": 1, "column": 2372 }]
                    },
                    {
                        "message": "Field \"user_created\" of type \"directus_users\" must have a selection of subfields. Did you mean \"user_created { ... }\"?",
                        "locations": [{ "line": 1, "column": 2411 }]
                    },
                    {
                        "message": "Field \"user_updated\" of type \"directus_users\" must have a selection of subfields. Did you mean \"user_updated { ... }\"?",
                        "locations": [{ "line": 1, "column": 2437 }]
                    },
                    {
                        "message": "Field \"user_created\" of type \"directus_users\" must have a selection of subfields. Did you mean \"user_created { ... }\"?",
                        "locations": [{ "line": 1, "column": 2544 }]
                    },
                    {
                        "message": "Cannot query field \"events\" on type \"kanban_test_m2a\".",
                        "locations": [{ "line": 1, "column": 2628 }]
                    }
                ],
                "stack": "DirectusError: GraphQL validation error.\n    at GraphQLService.execute (…/directus/api/src/services/graphql/index.ts:162:10)\n    at async <anonymous> (…/directus/api/src/controllers/graphql.ts:40:27)",
                "code": "GRAPHQL_VALIDATION"
            }
        }
    ]
}

```

</details>

paescuj committed:
Fixed GraphQL query generation

paescuj committed:
Revert queryToGqlString & sort imports

paescuj committed:
Update useItem test

paescuj reviewed (commented):
The GraphQL query needs to look as follows:
```gql
query {
  <collection>_by_id(id: <id>) {
    <...fields>
  }
}
```

For GraphQL we need to explicitly define all fields (including primary keys for relations), which means additional logic is required to fetch them for the query.
I've made the necessary changes in fd84f9628b735fe9692da5840b8f6e119b102f99.
Feel free to adjust further 🚀

paescuj committed:
Small cleanup

formfcw committed:
Merge branch 'main' into connorwinston/web-459-duplication-save-as-copy-of-a-multiple-relational-fields

formfcw reviewed (dismissed):
Thank you @paescuj for the great improvements to the PR! :heart:

However, it seems that the code is not able to handle M2A fields. Besides the comment in the code, I found out, that a bug occurs, when the value is cleared in the item duplication fields. When I try to duplicate an item, I get this graphql error:

```json
{
    "errors": [
        {
            "message": "GraphQL validation error.",
            "extensions": {
                "errors": [
                    {
                        "message": "Cannot query field \"group\" on type \"alignment\".",
                        "locations": [{ "line": 1, "column": 294 }]
                    }
                ],
                "stack": "DirectusError: GraphQL validation error.\n    at GraphQLService.execute (…/directus/api/src/services/graphql/index.ts:162:10)\n    at async <anonymous> (…/directus/api/src/controllers/graphql.ts:40:27)",
                "code": "GRAPHQL_VALIDATION"
            }
        }
    ]
}
```

… and a never-ending loading icon:

![progress](https://github.com/user-attachments/assets/afdce5ba-1966-4c19-b645-2ac199419be3)

connorwinston committed:
Merge branch 'main' into connorwinston/web-459-duplication-save-as-copy-of-a-multiple-relational-fields

paescuj committed:
Add initial m2a support to getGraphqlQueryFields

Not completely finished yet, need to map graphql response

ComfortablyCoding commented:
Should we account for export as well? See https://github.com/directus/directus/issues/23172

paescuj committed:
Always use GraphQL, full M2A support, other fixes

Other fixes meaning things that didn't work as expected before this PR
such as newly added rel. items that would have been lost before or
copying M2A that haven't been selected.
Tested various cases, should now be more stable overall.

paescuj committed:
Merge branch 'main' into connorwinston/web-459-duplication-save-as-copy-of-a-multiple-relational-fields

connorwinston committed:
Merge branch 'main' into connorwinston/web-459-duplication-save-as-copy-of-a-multiple-relational-fields

robluton committed:
Merge branch 'main' of github.com:directus/directus into connorwinston/web-459-duplication-save-as-copy-of-a-multiple-relational-fields

robluton committed:
show error dialog when saveAsCopy fails

formfcw committed:
Merge branch 'main' into connorwinston/web-459-duplication-save-as-copy-of-a-multiple-relational-fields

formfcw committed:
Merge branch 'main' into connorwinston/web-459-duplication-save-as-copy-of-a-multiple-relational-fields

formfcw committed:
Merge branch 'main' into connorwinston/web-459-duplication-save-as-copy-of-a-multiple-relational-fields

formfcw reviewed (changes requested):
Just one small change …

AlexGaillard committed:
Update app/src/composables/use-item/lib/get-graphql-query-fields.ts

Co-authored-by: Florian C. Wachmann <dev@formfcw.com>

formfcw reviewed (approved):
Thank you @connorwinston, great work @paescuj, also @robluton’s contribution to prevent the infinite loop error has improved this a lot.

I tested it with all fields and relationships and it …

LGTM :rocket:
